Add support for Analytics Engine

cmackenzie1 opened the pull request:
This one was a little tricky to add and requires mocking out the analytics engine binding using a miniflare wrapped binding that logs each data point to console.

Closes #280 
Closes #327

zebp commented on worker/src/analytics_engine.rs:
```diff
@@ -0,0 +1,186 @@
+        }
+    }
+
+    pub fn indexes(mut self, index: impl Into<Vec<String>>) -> Self {
```
For convenience, it'd be nice to have singular versions of all these methods that just added a single item to the inner vec.

zebp commented on worker/src/analytics_engine.rs:
```diff
@@ -0,0 +1,186 @@
+impl BlobType for Vec<u8> {}
+
+#[derive(Clone, Debug)]
+pub struct AnalyticsEngineDataPoint<T>
```
I think these should be type erased and under the hood the fields are JS arrays we can cheaply clone. This API as is feels a little painful always having to explicitly define a blob type that is only relevant when you're building.

zebp commented on worker/src/analytics_engine.rs:
```diff
@@ -0,0 +1,186 @@
+}
+
+pub struct AnalyticsEngineDataPointBuilder<T: BlobType> {
+    indexes: Vec<String>,
```
These should also just be JS Arrays we're settings/appending to.

zebp commented on worker/src/analytics_engine.rs:
```diff
@@ -0,0 +1,186 @@
+        self
+    }
+
+    pub fn blobs(mut self, blobs: impl Into<Vec<T>>) -> Self {
```
If we make this `impl Into<Vec<impl BlobType>>` we can prevent the BlobType generic from leaking into the other types.

zebp commented on worker/src/analytics_engine.rs:
```diff
@@ -0,0 +1,186 @@
+        self.blobs = blobs.into();
+        self
+    }
+
```
Also for convenience, a `write_to` method that accepts a `AnalyticsEngineDataset` so I can write it to WAE without having to then build and do a subsequent function call.

```rs
let ae = env.analytics_engine("whatever")?;
AnalyticsEngineDataPoint::builder()
    .doubles([1, 2, 3])
    .write_to(&ae)?;
```

zebp commented on worker/src/analytics_engine.rs:
```diff
@@ -0,0 +1,186 @@
+pub struct AnalyticsEngineDataPointBuilder<T: BlobType> {
+    indexes: Vec<String>,
+    doubles: Vec<f64>,
+    blobs: Vec<T>,
```
Notably the values in blobs can be an `Array<bytes | string>` not `Array<bytes> | Array<string>` so our builder methods can mix and match.

cmackenzie1 commented:
@zebp since the type is `Array<bytes | string>` I simplified the types to use an `enum` instead, removing all the generic parameters. Added the convenience function `write_to` as well as `doubles`, `blobs` which set the entire array and then `add_blob` and `add_double` which append to the current array of values.

zebp reviewed (approved):
LGTM other than some parameter types that I think could be less strict, going to merge into another branch so I can loosen the type requirements
